MatrixSDK/MXSession.m

```diff
@@ -1178,96 +1189,119 @@ - (void)serverSyncWithServerTimeout:(NSUInteger)serverTimeout
+    if (self->onBackgroundSyncFail)
+    {
+        NSLog(@"[MXSession] background Sync fails %@", error);
+
+        // Operations on session may occur during this block. For example, [MXSession close] may be triggered.
+        // We run a copy of the block to prevent app from crashing if the block is released by one of these operations.
+        MXOnBackgroundSyncFail onBackgroundSyncFailCpy = [self->onBackgroundSyncFail copy];
+        onBackgroundSyncFailCpy(error);
+        self->onBackgroundSyncFail = nil;
+
+        // check that the application was not resumed while catching up in background
+        if (self.state == MXSessionStateBackgroundSyncInProgress)
         {
-            NSLog(@"[MXSession] background Sync fails %@", error);
-            
-            // Operations on session may occur during this block. For example, [MXSession close] may be triggered.
-            // We run a copy of the block to prevent app from crashing if the block is released by one of these operations.
-            MXOnBackgroundSyncFail onBackgroundSyncFailCpy = [self->onBackgroundSyncFail copy];
-            onBackgroundSyncFailCpy(error);
-            self->onBackgroundSyncFail = nil;
-            
-            // check that the application was not resumed while catching up in background
-            if (self.state == MXSessionStateBackgroundSyncInProgress)
+            // Check that none required the session to keep running
+            if (self.preventPauseCount)
             {
-                // Check that none required the session to keep running
-                if (self.preventPauseCount)
-                {
-                    // Delay the pause by calling the reliable `pause` method.
-                    [self pause];
-                }
-                else
-                {
-                    NSLog(@"[MXSession] go to paused ");
-                    self->eventStreamRequest = nil;
-                    [self setState:MXSessionStatePaused];
-                    return;
-                }
+                // Delay the pause by calling the reliable `pause` method.
+                [self pause];
             }
             else
             {
-                NSLog(@"[MXSession] resume after a background Sync");
+                NSLog(@"[MXSession] go to paused ");
+                self->eventStreamRequest = nil;
+                [self setState:MXSessionStatePaused];
+                return;
             }
         }
-        
-        // Check whether the caller wants to handle error himself
-        if (failure)
+        else
         {
-            failure(error);
+            NSLog(@"[MXSession] resume after a background Sync");
+        }
+    }
+
+    // Check whether the caller wants to handle error himself
+    if (failure)
+    {
+        failure(error);
+    }
+    else
+    {
+        // Handle error here
+        // on 64 bits devices, the error codes are huge integers.
+        int32_t code = (int32_t)error.code;
```

Comment:
Check error domain before check error code. You could have an error from other domain with same error code.